atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/KvsSweepQueueTest.java

```diff
@@ -0,0 +1,243 @@
+
+import java.util.Map;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.palantir.atlasdb.encoding.PtBytes;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.api.Value;
+import com.palantir.atlasdb.keyvalue.impl.InMemoryKeyValueService;
+import com.palantir.atlasdb.schema.generated.TargetedSweepTableFactory;
+
+public class KvsSweepQueueTest {
+    private static final TableReference TABLE_CONSERVATIVE = TableReference.createFromFullyQualifiedName("test.cons");
+    private static final TableReference TABLE_THOROUGH = TableReference.createFromFullyQualifiedName("test.thor");
+    private static final TableReference TABLE_NOTHING = TableReference.createFromFullyQualifiedName("test.noth");
+    private static final Cell DEFAULT_CELL = Cell.create(new byte[] {'r'}, new byte[] {'c'});
+    private static final int CONS_SHARD = WriteInfo.tombstone(TABLE_CONSERVATIVE, DEFAULT_CELL, 0).toShard(SHARDS);
+    private static final int THOR_SHARD = WriteInfo.tombstone(TABLE_THOROUGH, DEFAULT_CELL, 0).toShard(SHARDS);
+    private static final long TS = 10L;
+    private static final long TS2 = 2 * TS;
+
+    KeyValueService kvs;
+    KvsSweepQueue sweepQueue = KvsSweepQueue.createUninitialized(() -> SHARDS);
+    KvsSweepQueueScrubber scrubber = mock(KvsSweepQueueScrubber.class);
+    long unreadableTs;
+    long immutableTs;
+
+    SweepTimestampProvider provider = new SweepTimestampProvider(() -> unreadableTs, () -> immutableTs);
+
+    @Before
+    public void setup() {
+        kvs = spy(new InMemoryKeyValueService(false));
+        unreadableTs = SweepQueueUtils.TS_COARSE_GRANULARITY;
+        immutableTs = SweepQueueUtils.TS_COARSE_GRANULARITY;
+        sweepQueue.initialize(provider, kvs);
+        sweepQueue.setScrubbers(scrubber);
+        kvs.createTable(TABLE_CONSERVATIVE, SweepQueueTestUtils.metadataBytes(CONSERVATIVE));
+        kvs.createTable(TABLE_THOROUGH, SweepQueueTestUtils.metadataBytes(THOROUGH));
+        kvs.createTable(TABLE_NOTHING, SweepQueueTestUtils.metadataBytes(NOTHING));
+    }
+
+    // todo(gmaretic): scrubber interface needs to be changed, because we also need to progress if there is nothing to sweep
+    @Test
+    public void scrubberCalledEvenIfNothingToSweep() {
+        sweepQueue.sweepNextBatch(ShardAndStrategy.conservative(CONS_SHARD));
+        verify(scrubber).scrub(anyCollection());
+    }
+
+    @Test
+    public void sweepStrategyNothingDoesNotPersistAntything() {
+        sweepQueue.enqueue(writeToDefaultCell(TABLE_NOTHING, TS), TS);
+        sweepQueue.enqueue(writeToDefaultCell(TABLE_NOTHING, TS2), TS2);
+        verify(kvs, times(0)).put(eq(TargetedSweepTableFactory.of().getSweepableCellsTable(null).getTableRef()),
```

Comment:
Looks like there should be no writes to the kvs. I haven't used spies before, but can you just verify no interactions instead of the individual checks here?

Reply:
We can do this